packages/wdio-config/src/lib/ConfigParser.ts

```diff
@@ -144,18 +140,31 @@ export default class ConfigParser {
      * Add hooks from an existing service to the runner config.
      * @param {Object} service - an object that contains hook methods.
      */
-    addService (service) {
-        for (let hookName of SUPPORTED_HOOKS) {
-            if (!service[hookName]) {
+    addService (service: Hooks) {
+        const addHook = <T extends keyof Hooks>(hookName: T, hook: Extract<Hooks[T], Function>) => {
+            const existingHooks: ConfigOptions[keyof Hooks] = this._config[hookName]
+            if (!existingHooks) {
+                // @ts-ignore Expression produces a union type that is too complex to represent
+                this._config[hookName] = hook.bind(service)
+            } else if (typeof existingHooks === 'function') {
+                this._config[hookName] = [existingHooks, hook.bind(service)]
+            } else {
+                this._config[hookName] = [...existingHooks, hook.bind(service)]
+            }
+        }
+
+        for (const hookName of SUPPORTED_HOOKS) {
+            const hooksToBeAdded = service[hookName]
+            if (!hooksToBeAdded) {
                 continue
             }
 
-            if (typeof service[hookName] === 'function') {
-                this._config[hookName].push(service[hookName].bind(service))
-            } else if (Array.isArray(service[hookName])) {
-                for (let hook of service[hookName]) {
-                    if (typeof hook === 'function') {
-                        this._config[hookName].push(hook.bind(service))
+            if (typeof hooksToBeAdded === 'function') {
+                addHook(hookName, hooksToBeAdded)
+            } else if (Array.isArray(hooksToBeAdded)) {
+                for (const hookToAdd of hooksToBeAdded) {
+                    if (typeof hookToAdd === 'function') {
+                        addHook(hookName, hookToAdd)
```

Comment:
How about we simplify this to:

```suggestion
                for (const hookToAdd of hooksToBeAdded.filter((hookToAdd) => typeof hookToAdd === 'function')) {
                    addHook(hookName, hookToAdd)
```

Reply:
The reason I did not simplify this to that, is that Typescript fails to narrow down the type of `hookToAdd` when used with a `filter()` function. You can check that in this [playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAFiEBrCAVEAhApgQQCZ5Z4BcMAFAGICuYwUAluDAD4zW0PgDaAugJS8YAXhi8A3ACgJAMxp1GYGAEMCACURIyCZADklAWyyloAJ3pgA5gBp4G9PhJs5nMHxgBvCTBgB6HzAA6IN8fCQBfKVBIWG0kPUNhGAByWIBaMAMsJMkZEBNyKOhbZHsCGBBpYpR0bAciN09vekqyKABPAAcsCqrSvGEhESTZDgUkhq9vZTUNLQ14rBtYvr5JbwiI3PyyQpi7EAdyyti0TFwCIgDpegAbKCwTMjmSg4I3IQA+GHaunuXX-qDIYjeTgcYTbwqPDqZDPOKZJb7ByrcISIA).

Reply:
We can assume here that `hooksToBeAdded` is from type `Function[]` which makes this work: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAFiEBrCAVEAhApgQQCZ5Z4BcMAFAGICuYwUAluDAD4zW0PgDaAugJS8YAXhi8A3ACgJAMxp1GYGAEMCACURIyCZADklAWyyloAJ3pgA5gBp4G9PhJs5nMHxgBvCTBgB6HzAA6IN8fCQBfKVBIWG0kPUNhGAByWIBaMAMsJMkZEBNyKOhbZHsCGBBpYpR0bAciN09vekqyKABPAAcsCqrSvGEhESTZDgUkhq9vZTUNLQ14rBtYvr5JbwiI3PyyQpi7EAdylti0TFwCImUIJ1HufgDpegAbKCwTMjmSg4I3IQA+GDtLo9ZbffqDIYjeTgcYTbwqPDqZCfOKZJb7ByrcISIA

Reply:
Okay, I'll make that change then.

Reply:
I'm getting the TS error when I'm trying to use `filter`:
```
This expression is not callable.
  Each member of the union type '{ <S extends (config: Config, capabilities: DesiredCapabilities[]) => void>(predicate: (value: (config: Config, capabilities: DesiredCapabilities[]) => void, index: number, array: ((config: Config, capabilities: DesiredCapabilities[]) => void)[]) => value is S, thisArg?: any): S[]; (predicate: (value: (config: Confi...' has signatures, but none of those signatures are compatible with each other.ts(2349)
```